ref(codeowners/notificationsettings): Add ExternalActor model

MeredithAnya opened the pull request:
**Context:**
The codeowners feature requires being able to map users and teams found in CODEOWNER files to their counterparts in Sentry. The initial architecture for this included two tables, `ExternalUser` and `ExternalTeam`, added in https://github.com/getsentry/sentry/pull/23698 and https://github.com/getsentry/sentry/pull/23696, respectively. 

The new notification settings feature work was going to use the `ExternalUser` as well to be able to store information needed for Slack notification preferences. 

Why not use the existing `Identity` table?
* For codeowners, we wanted to be able to have these mappings be created/updated/deleted via the API if need be, without requiring individual people to go through an oauth process. We don't need to do actions on _behalf_ of the user, so it's not necessary to have a token, or user permissions. These are essentially "unverified" external users, versus the Identity being "verified" external users (which may be needed for oauth based integrations to work). 
* For codeowners, we also wanted the external user to be scoped to the organization, which is why the original table has the `organization_member`. The identity table is not scope to the organization. 

**ExternalActor:**
The `Actor` table was added in https://github.com/getsentry/sentry/pull/24190, and seems to be the way we want to go moving forward. Basically the actor can be either a user or a team, and allows us to combine our tables into just one. Again this table will act as the "unverified" external users (when the actor is a user). 

In order to be flexible for both codeowners and notification settings (as well as any other features down the line) there are some changes from the prior two tables:

* `organization` - The `Actor` table assumes either user or team, and not the organization member. With the organization column we can implement checks to make sure that the user is a part of the organization we want. 
* `integration` - This is nullable because we don't _need_ it for codeowners, we want to allow users without a first party integration to still use this feature (similarly to how they use commits)
* `external_id` - Notification Settings will need this, but it may not be necessary for other features (like codeowners) 
* **`unique_together`** - we have the `organization`, `provider`, `external_name` and `actor` set here. 

**Other changes:**
I copied over the types file made in https://github.com/getsentry/sentry/pull/25049 and it seems like we don't really need the `ExternalProviderMixin` anymore.

Once this is merged we can:
- [ ] update the code to use this table instead of the `ExternalUser` and `ExternalTeam` tables
- [ ] delete any record from the `ExternalUser` and `ExternalTeam` tables (any data is dummy data, this is not live to users)
- [ ] remove the `ExternalUser` and `ExternalTeam` tables

mgaeta commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,62 @@
+class ExternalActor(DefaultFieldsModel, ExternalProviderMixin):
+    __core__ = False
+
+    actor = FlexibleForeignKey("sentry.Actor", null=True)
```
actor = FlexibleForeignKey(
    "sentry.Actor", db_index=True, unique=False, null=False, on_delete=models.CASCADE
)

mgaeta commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,62 @@
+    return EXTERNAL_PROVIDERS.get(ExternalProviders(value))
+
+
+class ExternalProviderMixin:
```
We had a mixin to duplicate logic for two classes. Now that theses are one model you can remove it.

mgaeta replied:
Actually we don't even need these two functions so you should probably just remove this class.

mgaeta commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,62 @@
+    organization = FlexibleForeignKey("sentry.Organization")
+    integration = FlexibleForeignKey("sentry.Integration", null=True)
+    provider = BoundedPositiveIntegerField(
+        choices=(
```
You should add PagerDuty now. We can come up with an ID for it and others now.

mgaeta commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,62 @@
+logger = logging.getLogger(__name__)
+
+
+class ExternalProviders(Enum):
```
I've moved this enum to a file in https://github.com/getsentry/sentry/pull/25046

mgaeta reviewed (changes requested):
I've left a few comments

MeredithAnya commented on src/sentry/migrations/0158_create_externalteam_table.py:
```diff
@@ -49,8 +50,8 @@ class Migration(migrations.Migration):
                         choices=[
-                            (sentry.models.integration.ExternalProviders(0), "github"),
-                            (sentry.models.integration.ExternalProviders(1), "gitlab"),
+                            (sentry.models.integration.ExternalProviders(200), "github"),
```
@wedamija I had to do this for my migration to be able to actually be generated. I don't know if this is okay to do. These tables (`ExternalTeam` and `ExternalUser`) will be dropped eventually, and I wanted the providers to be consistent in how they were numbered.

wedamija replied:
You shouldn't need to modify this migration to have the new migration generated - are you sure it wasn't just that the model hadn't been imported into `models/__init__.py`?

MeredithAnya commented on src/sentry/types/integrations.py:
```diff
@@ -3,17 +3,21 @@
     EMAIL = 100
     SLACK = 110
+    MSTEAMS = 120
+    PAGERDUTY = 130
```
Added `msteams` and `pagerduty` since those might be likely to come up soon

wedamija commented on src/sentry/migrations/0186_add_externalactor.py:
```diff
@@ -0,0 +1,74 @@
+
+
+class Migration(migrations.Migration):
+
```
Looks like our usual boilerplate with atomic/is_dangerous etc is missing, did you remove it?

MeredithAnya replied:
hmm that's weird I didn't remove it...at least not intentionally I can re-generate

MeredithAnya replied:
update for the masses: it was https://github.com/getsentry/sentry/pull/25236

MeredithAnya commented:
**Updated `ExternalProviders` Enum**

We previously had
```python
class ExternalProviders(Enum):
    GITHUB = 0 
    GITLAB = 1
```

I've updated the values to use the `100`, `110`, `120` etc number we do elsewhere BUT because we had the previous migrations expect `0` and `1` as valid values for the choices, I've renamed them so that we now have:
```python
class ExternalProviders(Enum):
    UNUSED_GH = 0 
    UNUSED_GL = 1
    ...
    GITHUB = 200
    GITLAB = 210
    ...
```

The human readable name shouldn't really matter here that we have two different values mapping to `"github"` because we are going to get rid of the `ExternalTeam` and `ExternalUser` altogether

mgaeta commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,36 @@
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_externalactor"
+        unique_together = (("organization", "provider", "external_name"),)
```
I think this will prevent a user from linking Slack usernames in multiple workspaces in an organizations. I think there's also a problem with using a nullable field in a `unique_together`.

MeredithAnya replied:
@mgaeta @ceorourke what if it was unique together with the actor_id as well? I could see the external_name being the same for different users, but I think it makes sense to limit a user to just having one workspace per org (like for workflow notifications for the user)

mgaeta commented on src/sentry/migrations/0186_add_externalactor.py:
```diff
@@ -0,0 +1,93 @@
+                    "provider",
+                    sentry.db.models.fields.bounded.BoundedPositiveIntegerField(
+                        choices=[
+                            (sentry.types.integrations.ExternalProviders(100), "email"),
```
You should paste a copy of the enum at the top of the file. That way this migration has no dependencies to the repo.

wedamija replied:
This is kind of a pain:
 - If we manually modify the migration to include the enum and use that here then (afaik) `makemigrations` will detect changes, since the enums aren't the same thing, so we can't really do it, unless someone knows of a workaround.
 - If we don't, then whenever we modify the enum it will modify all migrations that have ever used that enum.

I wonder if it's worth investigating https://pypi.org/project/django-enumfield/. We'd have to check whether this solves our problem.

ceorourke commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,36 @@
+
+    # external name => display name i.e. username, team name, channel name
+    # external id => unique identifier i.e user id, channel id
+    external_name = models.TextField()
```
maybe dumb question but why do we need both `external_name` and `external_id`? For notifications in Slack I believe we'll just need to store the IM channel ID, right? Is it for messaging teams in a channel so we'd need the channel ID and team name for the @?

MeredithAnya replied:
@ceorourke we need the GH/GL username or teamnames to be able to map the codeowners files. We won't have and don't need the `external_id` for that. It seems to me that if we have a slack ID we could easily get the name so I assumed it won't be a problem to have the external_name be non nullable but maybe that's not a fair assumption

ceorourke replied:
@MeredithAnya hmm yeah I'm trying to think through how we'll use this table (and make sure it works for both of our projects). For example I thought we could migrate data from the `Identity` table since we have some Sentry user IDs and Slack IDs already but we don't store the name there ... but maybe that can wait until we update permissions and we can do an API call to grab and store the name. That being said I'm not sure we need the name? Since like I said you DM a user by passing their user ID (no anchor tag unfortunately but look at "Post to a direct message channel
' in https://api.slack.com/methods/chat.postMessage#channels). What do you think @mgaeta ?

MeredithAnya commented:
Per outside discussion on some of the above comments:

* Added the `actor` to the `unique_together` constraint
* Keeping the `external_name` non-nullable, for cases like @ceorourke outlined where the name might not be necessarily to send the notification, we still want something to display to the user, and if it's not the actual "name" we can still show them the ID (so it could be save in both the `external_name` and `external_id` columns)

wedamija commented on src/sentry/models/externalactor.py:
```diff
@@ -0,0 +1,36 @@
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_externalactor"
+        unique_together = (("organization", "provider", "external_name", "actor"),)
```
Is there ever going to be a case where we have 2 slack integrations, that user has the same username in both workspaces and we want to have the user linked to both of these?

MeredithAnya replied:
@wedamija I think it's possible to have two slack integrations where the user has the same username for both workspaces, but I was under the impression that we didn't necessarily need to support that right now. 

From my understanding, we would have to do the following since the integration is nullable: 

```python
unique_together = (("organization", "provider", "external_name", "actor"), ("organization", "provider", "external_name", "actor", "integration"))
```

and that seemed fine to do later if we needed and we thought there was actually enough of use case for this. Looking at our data there are ~500 organizations that have more than 1 slack integration

@mgaeta @ceorourke  does that seem accurate to you both?

wedamija replied:
That's `unique_together` won't actually work - `("organization", "provider", "external_name", "actor")` will take precedence and prevent and duplicate rows on that key, so the less restrictive `unique_together` won't do anything.

You need partials, something like `CREATE UNIQUE INDEX ON <table> (organization_id, provider, external_name, actor_id) where integration_id is null` and `CREATE UNIQUE INDEX ON <table> (organization_id, provider, external_name, actor_id, integration_id) where integration_id is not null`.

Alternatively, if there was some way to encapsulate `provider, integration_id` in another table that would help, since you'd be removing the nulls from this table. I don't know if something along that lines makes sense though, just throwing out a general thought.

MeredithAnya replied:
so even though I was wrong about that `unique_together`, I think this still stands:

> I think it's possible to have two slack integrations where the user has the same username for both workspaces, but I was under the impression that we didn't necessarily need to support that right now.

We can deal with adding a constraint later when/if needed, but right now are okay with this limitation. Hopefully we can zoom boom boom to django 2.2 and use https://docs.djangoproject.com/en/2.2/ref/models/constraints/#django.db.models.UniqueConstraint.condition
